pep-0342.txt

```diff
@@ -5,591 +5,596 @@ Last-Modified: $Date$
-       iterator is garbage-collected.
-
-    6. Allow "yield" to be used in try/finally blocks, since garbage
-       collection or an explicit close() call would now allow the
-       finally clause to execute.
-
-    A prototype patch implementing all of these changes against the
-    current Python CVS HEAD is available as SourceForge patch #1223381
-    (http://python.org/sf/1223381).
+=====================
+
+By adding a few simple methods to the generator-iterator type, and with two
+minor syntax adjustments, Python developers will be able to use generator
+functions to implement co-routines and other forms of co-operative
+multitasking.  These methods and adjustments are:
+
+1. Redefine ``yield`` to be an expression, rather than a statement. The current
+   yield statement would become a yield expression whose value is thrown away.
+   A yield expression's value is ``None`` whenever the generator is resumed by
+   a normal ``next()`` call.
+
+2. Add a new ``send()`` method for generator-iterators, which resumes the
+   generator and *sends* a value that becomes the result of the current
+   yield-expression.  The ``send()`` method returns the next value yielded by
+   the generator, or raises ``StopIteration`` if the generator exits without
+   yielding another value.
+
+3. Add a new ``throw()`` method for generator-iterators, which raises an
+   exception at the point where the generator was paused, and which returns the
+   next value yielded by the generator, raising ``StopIteration`` if the
+   generator exits without yielding another value.  (If the generator does not
+   catch the passed-in exception, or raises a different exception, then that
+   exception propagates to the caller.)
+
+4. Add a ``close()`` method for generator-iterators, which raises
+   ``GeneratorExit`` at the point where the generator was paused.  If the
+   generator then raises ``StopIteration`` (by exiting normally, or due to
+   already being closed) or ``GeneratorExit`` (by not catching the exception),
+   ``close()`` returns to its caller.  If the generator yields a value, a
+   ``RuntimeError`` is raised.  If the generator raises any other exception, it
+   is propagated to the caller. ``close()`` does nothing if the generator has
+   already exited due to an exception or normal exit.
+
+5. Add support to ensure that ``close()`` is called when a generator iterator
+   is garbage-collected.
+
+6. Allow ``yield`` to be used in ``try/finally`` blocks, since garbage
+   collection or an explicit ``close()`` call would now allow the ``finally``
+   clause to execute.
+
+A prototype patch implementing all of these changes against the current Python
+CVS HEAD is available as SourceForge patch #1223381
+(http://python.org/sf/1223381).
 
 
 Specification: Sending Values into Generators
+=============================================
+
+New generator method: ``send(value)``
+
+   A new method for generator-iterators is proposed, called ``send()``.  It
```

Comment:
I just want to double-check that this indentation ends up looking okay in the final HTML?